[Aptos Framework] Add per-epoch stake increase/withdrawal limits

movekevin opened the pull request:
### Description
We want to limit the percentage of voting power increase and withdrawal per epoch to ensure the network's security and liveness.

### Test Plan
Unit tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/1745)
<!-- Reviewable:end -->

davidiw commented on aptos-move/framework/aptos-framework/sources/Genesis.move:
```diff
@@ -55,6 +59,10 @@ module AptosFramework::Genesis {
             allow_validator_set_change,
             rewards_rate,
             rewards_rate_denominator,
+            increase_limit,
+            increase_limit_denominator,
+            withdrawal_limit,
+            withdrawal_limit_denominator,
         )
```
we really need more than one init function or to pass in structs, this isn't really tenable :P

movekevin replied:
Doesn't creating the struct require the same number of args anyway? We can split this into 2 init functions but it's also rather awkward as all of these parameters are related. Plus the first init function would need to instantiate the empty ValidatorSetConfigurations resource with 0 for everything.

davidiw reviewed (commented):
I don't see anything obvious. I think the reality at this point in time, the entirety of the code needs a deep review. I'll delay approving in case someone else has more context and can provide a superior review.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -153,22 +167,46 @@ module AptosFramework::Stake {
+        // There are also potential different limits on how much voting power can be withdrawing or joining the
+        // validator set.
+        withdrawing: u128,
+        joining: u128,
```
should we unify on one word - joining or increase?

movekevin replied:
Updated to be consistently joining.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -460,11 +511,16 @@ module AptosFramework::Stake {
+        let validator_set_config = borrow_global<ValidatorSetConfiguration>(@CoreResources);
+        let validator_set_voting_power = borrow_global_mut<ValidatorSetVotingPower>(@CoreResources);
+        validator_set_voting_power.joining = validator_set_voting_power.joining + (amount_added as u128);
+        validate_voting_power_increase(validator_set_voting_power, validator_set_config);
```
this doesn't look right, we should only count the pools that are active or pending_active? if it's inactive, it seems double add to the joining field when the pool actually joins

movekevin replied:
Amazing catch!!! I updated the logic according to the suggestion and added unit tests to make sure we don't count inactive validator's stake in ValidatorSetVotingPower#joining.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -615,9 +674,14 @@ module AptosFramework::Stake {
+        let validator_set_voting_power = borrow_global_mut<ValidatorSetVotingPower>(@CoreResources);
+        validator_set_voting_power.withdrawing = validator_set_voting_power.withdrawing + (amount as u128);
+        let validator_set_config = borrow_global<ValidatorSetConfiguration>(@CoreResources);
+        validate_voting_power_withdrawal(validator_set_voting_power, validator_set_config);
```
here too, we should not count pools not in the validator set

movekevin replied:
Updated the logic according to the suggestion and added unit tests to make sure we don't count inactive validator's stake in ValidatorSetVotingPower#leaving.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -786,16 +864,31 @@ module AptosFramework::Stake {
         let active_validators = Vector::empty();
         validator_perf.num_blocks = 1;
         validator_perf.missed_votes = Vector::empty();
+
```
curious why line 865 the num_blocks is 1 instead of 0?

movekevin replied:
Good catch! It should be 0. on_new_epoch() is called after update_performance_statistics (which increments the number of blocks) in Block#prologue. So this should start out as 0 for a new epoch but immediately incremented to 1 once update_performance_statistics is called.

movekevin replied:
Good catch! This should be 0. I was thinking 1 because the rewards distribution in on_new_epoch would fail with division by 0 - but this should not happen outside of tests as update_performance_statistics is always called after on_new_epoch in Block#prologue.

movekevin replied:
Updated to be 0 now + updated tests correspondingly.

zekun000 replied:
now we know that we need to call update_performance_statistics before on_new_epoch 😄

movekevin replied:
Yes it makes more sense now.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -865,7 +958,22 @@ module AptosFramework::Stake {
+        // power of a validator cannot exceed the max allowed (the excess stake above the max is effectively idle).
+        // Second, this prevents a potential manipulation where the validator withdraws the excess stake in order to
+        // receive rewards on both active and pending_inactive as the amounts individually do not exceed the max.
+        if (Coin::value(&stake_pool.active) + Coin::value(&stake_pool.pending_inactive) <= validator_set_config.maximum_stake) {
```
hmm, this seems counter-intuitive, I thought they only receive rewards for the maximum stake regardless of how much is being withdrawn

movekevin replied:
This does exactly that. It's just an implementation details - we distribute the rewards for active and pending_inactive separately (since those are 2 different Coin objects).

movekevin replied:
I moved this calculation out to before the reward distribution and named it total_stake_before_rewards to be clearer.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -470,30 +521,38 @@ module AptosFramework::Stake {
         if (is_current_validator(pool_address)) {
             Coin::merge<TestCoin>(&mut stake_pool.pending_active, coins);
+
+            // Update the amount of stake being added to the active validator set's total voting power.
```
I think there's still some nuance here, is_current_validator includes the pending_inactive nodes too.

if the node is in the pending_active queue (not current validator), we should count the added value as well

movekevin replied:
Updated

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -672,6 +735,14 @@ module AptosFramework::Stake {
         if (is_current_validator(pool_address)) {
             Coin::merge<TestCoin>(&mut stake_pool.pending_inactive, unlocked_stake);
+
+            // Update the amount of stake being withdrawn from the active validator set's total voting power.
```
if it's already a pending_inactive validator, the voting power is being counted twice

movekevin replied:
Updated

zekun000 reviewed (commented):
we need to enumerate all cases of validators and stakes to make it right
validators have four states - inactive, pending_active, active, pending_inactive.
stakes have those four states as well.

some combination will never happen (need to double check) e.g. inactive/pending_active validator does not have any pending stakes.

movekevin commented:
> we need to enumerate all cases of validators and stakes to make it right validators have four states - inactive, pending_active, active, pending_inactive. stakes have those four states as well.
> 
> some combination will never happen (need to double check) e.g. inactive/pending_active validator does not have any pending stakes.

Thanks for catching these cases. I updated the code to take them all into account and added unit tests to cover all cases.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -623,14 +692,20 @@ module AptosFramework::Stake {
         validate_lockup_time(stake_pool.locked_until_secs, validator_set_config);
 
-        // Throw an error is the validator is already active.
+        // Throw an error is the validator is already active, pending_active, or pending_inactive.
```
the function name `is_validator_active` is confusing, probably should be `is_validator_inactive` and negates the boolean

movekevin replied:
Why is it confusing?

movekevin replied:
This will be changed in https://github.com/aptos-labs/aptos-core/pull/1860

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -668,6 +743,22 @@ module AptosFramework::Stake {
 
+        // Update the amount of stake being withdrawn from the active validator set's total voting power.
+        // We only do this if the validator is active.
+        // If the validator is pending_active, it's not possible to unlock immediately if we assume that the min lockup
```
I think it's better to verify than assume, when we update min lockup or epoch duration we should check the invariant if we're going down this path. or explicitly assert this pool is not in pending_active vector

movekevin replied:
It's really tricky to check epoch length in Stake as Block already depends on Stake (via Reconfiguration). So if Stake also calls Block to get the epoch length, there'd be a circular dependency. We can technically move the epoch length out of Block into another module, but that sounds like adding a lot more code.

zekun000 replied:
we can explicitly prevent withdraw if it's in pending_active then? I'm always skeptical of implicit invariants that may potentially break

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -828,18 +932,33 @@ module AptosFramework::Stake {
+            };
+
+            // A validator needs at least the min stake required to join the validator set.
+            if (new_validator_info.voting_power >= validator_set_config.minimum_stake) {
```
this is another edge case not sure if we want to handle - if someone withdraw only 10 but resulted in the node below the minimum_stake, the actual reduction is going to be the all the voting power from that pool.

zekun000 replied:
the easiest thing to do is probably just prevent people from withdrawing below minimum_stake. they have to leave first

movekevin replied:
Updated to prevent active validators (not considering pending_inactive) from unlocking if their remaining stake is less than min stake required. I also added a unit test just for that case.

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -668,6 +743,28 @@ module AptosFramework::Stake {
+        let validator_set_voting_power = borrow_global_mut<ValidatorSetVotingPower>(@CoreResources);
+        if (Option::is_some(&find_validator(&validator_set.active_validators, pool_address))) {
+            // If the validator is still active, we do not allow unlocking if remaining stake is below
+            // the min stake required. The validator would need to leave the validator set first.
```
we have to make sure the minimum stake is always <= withdraw limit otherwise all of them are deadlocked

zekun000 commented on aptos-move/framework/aptos-framework/sources/configs/Stake.move:
```diff
@@ -908,8 +1033,25 @@ module AptosFramework::Stake {
         };
 
+        let total_stake_before_rewards =
+            Coin::value(&stake_pool.active) + Coin::value(&stake_pool.pending_inactive);
```
i'm now paranoid if the stake + reward overflows, but that's orthogonal to this PR

movekevin replied:
Yeah I have a task to revisit the balance type for Coin. u64 only covers up to ~20**19. If the coin has 10 decimals, initial supply of 1B coins is already 10^19 which is going to be very close to the the limit already.
